File: Sources/waistgunners/Bullet.swift
import Foundation
import swallegro

let BULLET_SIZE          = Vector(x : 4, y : 4)
let BULLET_SPEED : Float = 3

class Bullet : Entity {
    var active : Bool = true
    var motion : Vector
    var player : Bool
    var lifetime : Int = 0

    init(pos : Vector, motion : Vector, player : Bool) {
        self.motion = motion
        self.player = player
        super.init(pos : pos)
        self.size = BULLET_SIZE
    }

    func stop() {
        self.active = false
    }

    override func update() {
        self.position = self.motion * BULLET_SPEED + self.position
        self.lifetime = self.lifetime + 1
        if self.position.x < 0 || self.position.y < 0 ||
            Int(self.position.x) > buffer_width || Int(self.position.y) > buffer_height {
            self.active = false
        }
    }

    override func draw() {
        var sprite : SPRITES
        if self.player {
            if self.lifetime % 50 > 25 {
                sprite = SPRITE_PLAYER_BULLET_1
            } else {
                sprite = SPRITE_PLAYER_BULLET_2
            }
        } else {
            if self.lifetime % 50 > 25 {
                sprite = SPRITE_ENEMY_BULLET_1
            } else {
                sprite = SPRITE_ENEMY_BULLET_2
            }
        }
        draw_bitmap(sprite, position : self.position.apply(function : {(_ input : Vector) in
            return Vector(x : Float(Int(input.x)), y : Float(Int(input.y)))
        }), flags : 0)
    }
}

File: scripts/Bullet.js
const BULLET_SPEED = 3;
const BULLET_SIZE  = new Vector(3, 3);

class Bullet extends Entity {
    constructor(position, motion, owner) {
        super(position);
        this.owner    = owner;
        this.active   = true;
        this.motion   = motion;
        this.lifetime = 0;
    }

    update() {
        this.lifetime++;
        this.position = this.position.plus(this.motion.times(BULLET_SPEED));
        this.active   = (
            this.position.x > 0 && this.position.x < BUFFER_WIDTH &&
            this.position.y > 0 && this.position.y < BUFFER_HEIGHT
        ) && this.active;
        return this;
    }

    hit() {
        this.active = false;
    }

    draw() {
        var sprite;
        if (this.owner == "player") {
            if (this.lifetime % 35 > 17) {
                sprite = sprites.player_bullet_1;
            } else {
                sprite = sprites.player_bullet_2;
            }
        } else if (this.owner == "enemy") {
            if (this.lifetime % 20 > 10) {
                sprite = sprites.enemy_bullet_1;
            } else {
                sprite = sprites.enemy_bullet_2;
            }
        }
        context.drawImage(sprite, this.position.x, this.position.y);
    }
}